waspc/src/Wasp/Generator/ServerGenerator/OperationsG.hs

```diff
@@ -10,45 +10,95 @@ where
-import Data.Maybe (fromJust)
+import Data.List (nub)
+import Data.Maybe (fromJust, fromMaybe)
 import StrongPath (Dir, Dir', File', Path, Path', Posix, Rel, relDirToPosix, reldir, reldirP, relfile, (</>))
 import qualified StrongPath as SP
 import Wasp.AppSpec (AppSpec)
 import qualified Wasp.AppSpec as AS
 import qualified Wasp.AppSpec.Action as AS.Action
+import Wasp.AppSpec.Operation (getName)
 import qualified Wasp.AppSpec.Operation as AS.Operation
 import qualified Wasp.AppSpec.Query as AS.Query
+import Wasp.AppSpec.Valid (isAuthEnabled)
 import Wasp.Generator.ExternalCodeGenerator.Common (GeneratedExternalCodeDir)
 import Wasp.Generator.FileDraft (FileDraft)
 import Wasp.Generator.JsImport (getJsImportDetailsForExtFnImport)
 import Wasp.Generator.Monad (Generator)
 import qualified Wasp.Generator.ServerGenerator.Common as C
 import Wasp.Generator.ServerGenerator.ExternalCodeGenerator (extServerCodeDirInServerSrcDir)
-import Wasp.Util ((<++>))
+import Wasp.Util (toUpperFirst, (<++>))
 
 genOperations :: AppSpec -> Generator [FileDraft]
-genOperations spec = genQueries spec <++> genActions spec
+genOperations spec =
+  sequence
+    [ genQueryTypesFile spec,
+      genActionTypesFile spec
+    ]
+    <++> genQueries spec
+    <++> genActions spec
 
 genQueries :: AppSpec -> Generator [FileDraft]
-genQueries spec = mapM (genQuery spec) (AS.getQueries spec)
+genQueries spec = mapM genQuery (AS.getQueries spec)
 
 genActions :: AppSpec -> Generator [FileDraft]
-genActions spec = mapM (genAction spec) (AS.getActions spec)
+genActions spec = mapM genAction (AS.getActions spec)
+
+genQueryTypesFile :: AppSpec -> Generator FileDraft
+genQueryTypesFile spec = genOperationTypesFile tmplFile dstFile operations isAuthEnabledGlobally
+  where
+    tmplFile = [relfile|src/queries/types.ts|]
+    dstFile = [relfile|src/queries/types.ts|]
+    operations = map (uncurry AS.Operation.QueryOp) $ AS.getQueries spec
+    isAuthEnabledGlobally = isAuthEnabled spec
+
+genActionTypesFile :: AppSpec -> Generator FileDraft
+genActionTypesFile spec = genOperationTypesFile tmplFile dstFile operations isAuthEnabledGlobally
+  where
+    tmplFile = [relfile|src/actions/types.ts|]
+    dstFile = [relfile|src/actions/types.ts|]
+    operations = map (uncurry AS.Operation.ActionOp) $ AS.getActions spec
+    isAuthEnabledGlobally = isAuthEnabled spec
 
 -- | Here we generate JS file that basically imports JS query function provided by user,
 --   decorates it (mostly injects stuff into it) and exports. Idea is that the rest of the server,
 --   and user also, should use this new JS function, and not the old one directly.
-genQuery :: AppSpec -> (String, AS.Query.Query) -> Generator FileDraft
-genQuery _ (queryName, query) = return $ C.mkTmplFdWithDstAndData tmplFile dstFile (Just tmplData)
+genQuery :: (String, AS.Query.Query) -> Generator FileDraft
+genQuery (queryName, query) = return $ C.mkTmplFdWithDstAndData tmplFile dstFile (Just tmplData)
```

Comment:
@shayneczyzewski I removed `AppSpec` from the list of arguments because the function doesn't use it.

I know you somtimes keep unused arguments for the sake of API consistency, but in this case I only see two functions that "belong" to the API (`genQuery` and `genAction`) and neither uses the `AppSpec` argument. Correct me if I missed something.

Reply:
Sounds good to me! I think those signatures have had the param ignored for a few years now, so fully agree cleanup makes sense if unused. 👍🏻

Reply:
Nooooooo :D. Kidding, and I wonder if we should have at some point put AppSpec in Generator Monad. But maybe not. Anyway all good.